### PR TITLE
Reader alembic uv normal support

### PR DESCRIPTION
**Pull request by mwestphal**

### CODE REVIEW

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 1]
```diff
@@ -20,51 +22,330 @@
...
+    if (need_to_duplicate)
+    {
+      auto face_count = original_data._indices.size();
+      duplicated_data._indices.resize(face_count);
+      for (size_t i = 0; i < face_count; i++)
       {
-        points->SetPoint(i, positions->get()[i].x, positions->get()[i].y, positions->get()[i].z);
+        auto this_face_vertex_count = original_data._indices[i].size();
+        duplicated_data._indices[i].resize(
+          this_face_vertex_count, Alembic::Abc::V3i(-9999, -9999, -9999));
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:105]: what does this -9999 value correspond to ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 2]
```diff
@@ -20,51 +22,330 @@
...
+
+  void FillPolyData(const IntermediateGeometry& data, vtkPolyData* polydata)
+  {
+    // Create 10 points.
+    vtkNew<vtkPoints> points;
+    vtkNew<vtkCellArray> cells;
 
-      polys->SetData(offsets, connectivity);
+    auto P_map_iter = data._attributes.find("P");
+    assert(P_map_iter != data._attributes.end());
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:185]: @nyue can this actually happen ?

**Comment by mwestphal** [comment]
@Meakk @nyue please review

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 3]
```diff
@@ -20,51 +22,338 @@
...
+    {
+      auto faceCount = originalData.Indices.size();
+      duplicatedData.Indices.resize(faceCount);
+      for (size_t i = 0; i < faceCount; i++)
+      {
+        auto thisFaceVertexCount = originalData.Indices[i].size();
+
+        // TODO check with nYue
+        duplicatedData.Indices[i].resize(
+          thisFaceVertexCount, Alembic::Abc::V3i(-9999, -9999, -9999));
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:107]: @nyue what does -9999 means ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 4]
```diff
@@ -20,51 +22,338 @@
...
+    vtkNew<vtkPoints> points;
+    vtkNew<vtkCellArray> cells;
+
+    auto pMapIter = data.Attributes.find("P");
+    if (pMapIter == data.Attributes.end())
+    {
+      // TODO throw an error instead and catch it above ?
+      vtkWarningWithObjectMacro(nullptr, "A data could not be read, leaving empty");
+      return;
+    }
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:192]: A warning here seems fine, wdyt @nyue ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 5]
```diff
@@ -20,51 +22,338 @@
...
+          }
+        }
+
+        duplicatedData.Attributes.insert(AttributesContainer::value_type("N", nV3F));
+      }
+    }
+    else
+    {
+      duplicatedData = originalData;
+    }
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:177]: is there data somewhere that would test this code path @nyue ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 6]
```diff
@@ -20,51 +22,338 @@
...
+      vtkNew<vtkFloatArray> normals;
+      normals->SetNumberOfComponents(3);
+      for (auto& N : nMapIter->second)
+      {
+        normals->InsertNextTuple3(N.x, N.y, N.z);
+      }
+      if (data.nIsFaceVarying)
+      {
+        cellAttributes->SetNormals(normals);
+      }
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:231]: do you have nFaceVarying data to test this with @nyue ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 7]
```diff
@@ -20,51 +22,338 @@
...
+          {
+            originalData.uvFaceVarying = true;
+            this->UpdateIndices<Alembic::AbcGeom::UInt32ArraySamplePtr>(
+              uvIndices, uvIndicesOffset, originalData.Indices);
+          }
+          else
+          {
+            this->UpdateIndices<Alembic::AbcGeom::Int32ArraySamplePtr>(
+              face_positionIndices, uvIndicesOffset, originalData.Indices);
+          }
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:316]: do you have non-kFacevaryingScope data to test this with @nyue ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 3 above)
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:107]: I was originally thinking that initialising it with some known sentinel value makes it easier to debug and when I see -9999, I know I have miscalculated what needs to be initialise with proper value down the line. Feel free to change it if this defensive programming approach does not have sufficient merit.

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 5 above)
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:177]: [tetrahedron.zip](https://github.com/f3d-app/f3d/files/13396844/tetrahedron.zip)

I have tested it with the above a satisfy your testing requirements.

It's a tetrahedron. Both the UV and Normal are point attributes i.e. all non-facevarying

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 4 above)
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:192]: Having no "P" is still a valid Alembic though it makes no sense in a geometry viewer. For example, Alembic camera will have no "P" which is a valid use case but for f3d not really (unless there are plans to support camera)

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 6 above)
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:231]: [tetrahedron_facevarying_N.zip](https://github.com/f3d-app/f3d/files/13398760/tetrahedron_facevarying_N.zip)

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 8]
```diff
@@ -20,51 +22,338 @@
...
+      vtkNew<vtkFloatArray> uvs;
+      uvs->SetNumberOfComponents(2);
+      for (auto& uv : uvMapIter->second)
+      {
+        uvs->InsertNextTuple2(uv.x, uv.y);
+      }
+      if (data.uvFaceVarying)
+      {
+        cellAttributes->SetTCoords(uvs);
+      }
```
[the remark this answers is not in the document]
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:249]: The suzanne.abc is already containing facevarying UVs

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 7 above)
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:316]: I believe the first zip I attached in the comments above may fit this test requirements. Let me know if it is another corner case I need to generate specific geometry file for.

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 4 above)
Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:192]: Then we should not warn at all.

> Unless there are plans to support camera

We do support reading cameras but only for importers. FYI @Meakk

Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:192]: could you share such a file, containing camera I mean @nyue ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 8 above)
[the remark this answers is not in the document]
Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:249]: I'm afraid this specific code is not reached when reading `suzanne.abc` .

`if (data.uvFaceVarying)` is not true.

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 6 above)
Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:231]: I'm afraid opening above file with F3D does not reach this code. Is that expected ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 4 above)
Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:188]: [alembic_camera.zip](https://github.com/f3d-app/f3d/files/13401092/alembic_camera.zip)

Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:188]: This file does not reach this code because there is neither `Alembic::AbcGeom::IXform::matches` nor `Alembic::AbcGeom::IPolyMesh::matches`.

So maybe this code is unreachable after all.

Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:188]: The alembic_camera.zip contains an alembic file purely with camera information, no geometry at all i.e. no "P", no polymesh.

Comment (Meakk) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:190]: Let's drop camera support for now.

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 9]
```diff
@@ -20,51 +22,319 @@
...
+    if (schema.getNumSamples() > 0)
     {
-      pmesh.getSchema().get(samp);
+
+      schema.get(samp);
 
       Alembic::AbcGeom::P3fArraySamplePtr positions = samp.getPositions();
-      Alembic::AbcGeom::Int32ArraySamplePtr indices = samp.getFaceIndices();
-      Alembic::AbcGeom::Int32ArraySamplePtr counts = samp.getFaceCounts();
+      Alembic::AbcGeom::Int32ArraySamplePtr face_positionIndices = samp.getFaceIndices();
```
Comment (Meakk) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:252]: snake case -> camel case

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 10]
```diff
@@ -20,51 +22,319 @@
...
+      {
+        V3fContainer nV3F;
+        int nRunningIndex = 0;
+
+        for (size_t i = 0; i < faceCount; i++)
+        {
+          auto thisFaceVertexCount = originalData.Indices[i].size();
+          for (size_t j = 0; j < thisFaceVertexCount; j++)
+          {
+            Alembic::Abc::V3f original_N =
```
Comment (Meakk) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:162]: camel case

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 4 above)
Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:190]: Of course, we have no way to support camera. I'm just trying to improve coverage here.

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 11]
```diff
@@ -20,51 +22,319 @@
...
+      auto faceCount = originalData.Indices.size();
+      duplicatedData.Indices.resize(faceCount);
+      for (size_t i = 0; i < faceCount; i++)
+      {
+        auto thisFaceVertexCount = originalData.Indices[i].size();
+        duplicatedData.Indices[i].resize(thisFaceVertexCount, Alembic::Abc::V3i());
+      }
+
+      // Points
+      {
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:110]: it does deallocate the `V3fContainer` variable, which could be reducing memory usage of this method.

@nyue may know more.

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 12]
```diff
@@ -20,51 +22,319 @@
...
-      Alembic::AbcGeom::Int32ArraySamplePtr counts = samp.getFaceCounts();
+      Alembic::AbcGeom::Int32ArraySamplePtr face_positionIndices = samp.getFaceIndices();
+      Alembic::AbcGeom::Int32ArraySamplePtr faceVertexCounts = samp.getFaceCounts();
 
-      points->SetNumberOfPoints(positions->size());
+      this->SetupIndicesStorage(faceVertexCounts, originalData.Indices);
 
-      for (size_t i = 0; i < positions->size(); i++)
+      // Positions
       {
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:258]: same.

@nyue

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 11 above)
[the remark this answers is not in the document]
Comment (Meakk) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:109]: Ok, I guess that's fine then.

File: testing/baselines/TestABCFaceVarying.png
[the remark this answers is not in the document]
Comment (Meakk) [reply testing/baselines/TestABCFaceVarying.png:1]: Looks like suzanne.abc is face-varying, I can see the quad having the same normals, as opposed to before where we could clearly see the separating edge of the quad.

**Review by Meakk — APPROVED** [review]
